[core][autoscaler][v2] Instance manager: Instance Utils [1/x]

rickyyx opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

InstanceUtil to 
- wrap around the Instance proto message
- explicitly encode instance status transistion dag.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks
```
---------- coverage: platform linux, python 3.9.16-final-0 -----------
Name                                                            Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------------------
python/ray/autoscaler/v2/instance_manager/instance_manager.py      54      3    94%   53, 68, 149
---------------------------------------------------------------------------------------------
TOTAL                                                              54      3    94%


Results (0.07s):
       4 passed
```
- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,215 @@ def get_instance_manager_state(
+        there's already a cloud node for the instance.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in {
```
Make sure we have test to make sure all the instance statuses (include future statuses) are covered.

rickyyx replied:
good point

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,215 @@ def get_instance_manager_state(
+    @staticmethod
+    def is_ray_pending(instance: Instance) -> bool:
+        """
+        Returns True if the instance is pending ray installation or ray startup,
+        regardless of whether the instance is allocated by the cloud provider.
```
Change the comment saying that it should include all states that can be transitioned into RAY_RUNNING.

Rename the function as well to reflect that.

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,215 @@ def get_instance_manager_state(
+            instance: The instance to update.
+            status: The new status to transition to.
+        """
+        now_ms = time.time_ns() // 1000000
```
Use the monotonically increasing timer.

Or record the wall clock time once at the beginning and then record elapsed time.

rickyyx replied:
Decided to just not enforcing the ordered timestamp since it's purely observability features.

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,215 @@ def get_instance_manager_state(
+            )
+
+    @staticmethod
+    def get_status_time_ms(
```
get_status_last_time_ms

rickyyx commented:
- Added explicit graphs tests
- Added tests for iterating through all possible states to catch future regression
- Remove constrain on the ordering of status history 
- Modify get_status_time_ms API.

rkooo567 commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+        instance.launch_request_id = request_id
+        instance.total_resources.update(resources)
+        instance.status = Instance.QUEUED
+        InstanceUtil._record_status_transition(
```
_record_status_transition should be a public method of `instance` instead?

rickyyx replied:
it's expected to use within `set_status` for now.

rkooo567 commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+        ]
+
+    @staticmethod
+    def set_status(
```
Q: maybe I am missing some context, but why are these a part of util method not a public API of instance class?

rickyyx replied:
Instance is a proto message so..

rkooo567 commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+        to RAY_RUNNING status.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in [
```
should this be dynamically calculated based on get_valid_transitions? We can calculate once and cache the result

rkooo567 replied:
(or if you plan to keep this guarantee via unit test, that sounds good to me too)

rickyyx replied:
it's kept in unit testing now.

rkooo567 commented on python/ray/autoscaler/v2/tests/test_instance_manager.py:
```diff
@@ -0,0 +1,172 @@
+
+
+class InstanceUtilTest(unittest.TestCase):
+    def test_transition_graph(self):
```
add a test to verify is_ray_running_reachable is correct based on the graph?

rickyyx replied:
yes - it's in `test_is_ray_running_reachable`

rkooo567 reviewed (commented):
looking at util method, it looks like most of them could be just a part of instance class. Is there a reason why we are using static method util here?

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+            instance: The instance to update.
+            status: The new status to transition to.
+        """
+        now_ms = time.time_ns() // 1000000
```
Recommend to store ns and during display, you can show ms if you want.

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+                # Ray needs to be install and launch on the provisioned cloud node.
+                Instance.RAY_INSTALLING,
+                # Ray is already installed and running on the provisioned cloud node.
+                Instance.RAY_RUNNING,
```
Could you add comments about when each transition can happen? E.g. when will we transition from allocated directly to ray_running.

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+                Instance.RAY_RUNNING,
+                # Instance is requested to be stopped, e.g. instance leaked.
+                Instance.STOPPING,
+                # Cloud node somehow failed.
```
is it necessarily failure?

rickyyx replied:
I think for ALLOCATED -> STOPPED, without going through the STOPPING status, it will be failure.

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
+        }
+
+    @staticmethod
+    def get_status_time_ms(
```
recommend to return ns

jjyao replied:
nit: get_status_update_times?

plural

rickyyx commented:
- Added state machine transition comments 
- Added InvalidInstanceStatusError for invalid instance status instead of generic ValueError
- Rename get_status_times_ns to better match the behaviour

jjyao commented on python/ray/autoscaler/v2/instance_manager/instance_manager.py:
```diff
@@ -63,3 +67,252 @@ def get_instance_manager_state(
+                # is discovered in the set of running ray nodes from the Ray cluster.
+                Instance.RAY_RUNNING,
+                # Instance is requested to be stopped, e.g. instance leaked.
+                # This happens when the autoscaler is terminating the ray process on
+                # the instance, e.g. idle termination.
```
This should be RAY_STOPPING? I feel this comment is not right.

rickyyx replied:
yeah, good call

jjyao reviewed (commented):
lgtm
